File: 2024/dart/lib/aoc.dart
import 'dart:io' as io;

String read_input(String file_path) {
    final file = io.File(file_path);
    try {
        return file.readAsStringSync();
    } catch (e) {
        print('Error reading file: $e');
        throw e;
    }
}

typedef Solution = void Function(String input);

void run(String file_path, Solution solution) {
    final file = read_input(file_path);
    print("==== $file_path");
    final start = DateTime.now();
    solution(file);
    final elapsed = (DateTime.now().microsecondsSinceEpoch - start.microsecondsSinceEpoch).toDouble() / 1000;
    print("==== time: ${elapsed}ms");
}

File: 2024/rust/src/aoc.rs
fn read_input(path: &str) -> String {
    return std::fs::read_to_string(path).unwrap();
}

type Solution = fn(input: &str);

pub fn run(path: &str, solution: Solution) {
    println!("==== {}", path);
    let start_time = std::time::Instant::now();
    solution(&read_input(path));
    println!("==== time: {:?}", start_time.elapsed());
}

